README.md
[![Build Status](https://travis-ci.org/korzio/djv.svg?branch=master)](https://travis-ci.org/korzio/djv)
[![Join the chat at https://gitter.im/korzio/djv](https://badges.gitter.im/korzio/djv.svg)](https://gitter.im/korzio/djv)

# djv <a name="title"></a>

Dynamic JSON-Schema Validator

[Official Documentation](http://korzio.github.io/djv/)

Current package supports **JSON-Schema v6 and v4**. It contains utils to validate objects against schemas.
This is a part of **djv** packages aimed to work with json-schema.

- [djv](https://github.com/korzio/djv) validate object against schemas
- [djvi](https://github.com/korzio/djvi) instantiate objects by schema definition
- [jvu](https://github.com/korzio/jvu) utilities for declarative, FP usage
- [@djv/draft-04](https://github.com/korzio/@djv/draft-04) environment updates to support *draft-04*

Any contributions are welcome. Check the [contribution guide](./CONTRIBUTING.md).
Since version **1.2.0** *djv* package supports `draft-06`. Version **2.0.0** makes `draft-06` the default schema version. To use other versions check the [environment section](#environment).

## Table of contents <a name="content"></a>

* [djv](#title)
* [Table of contents](#content)
* [Install](#install)
* [Usage](#usage)
* [API](#api)
  * [Environment](#environment)
  * [addSchema](#addSchema)
  * [validate](#validate)
  * [removeSchema](#removeSchema)
  * [resolve](#resolve)
  * [export](#export)
  * [import](#import)
  * [addFormat](#addFormat)
  * [setErrorHandler](#errorHandler)
  * [useVersion](#useVersion)
* [Tests](#tests)
* [References](#references)

## Install <a name="install"></a>

```bash
npm install djv
```

or

```html
<script src="djv.js"></script>
```

There are 2 versions of validator

- `./lib/djv.js` a default one, not uglified and not transpiled
- `./djv.js` a built one with a webpack, babel and uglify (preferable for frontend)

## Usage <a name="usage"></a>

```javascript
const env = new djv();
const jsonSchema = {
  "common": {
    "properties": {
      "type": {
        "enum": ["common"]
      }
    },
    "required": [
      "type"
    ]
  }
};

// Use `addSchema` to add json-schema
env.addSchema('test', jsonSchema);
env.validate('test#/common', { type: 'common' });
// => undefined

env.validate('test#/common', { type: 'custom' });
// => 'required: data'
```

## API <a name="api"></a>

### Environment <a name="environment"></a>

To instantiate *djv* environment

```javascript
const djv = require('djv');
const env = djv({
  version: 'draft-06', // use json-schema draft-06
  formats: { /*...*/ }, // custom formats @see #addFormat
  errorHandler: () => { /*...*/ }, // custom error handler, @see #setErrorHandler
});
```

To use a previous version of JSON-Schema draft, use a [`draft-04` plugin](https://www.npmjs.com/package/@korzio/djv-draft-04), specified in [*optionalDependencies*](https://docs.npmjs.com/files/package.json#optionaldependencies) of *djv* package.

```javascript
const env = new djv({ version: 'draft-04' });
```

### addSchema(name: string, schema: object?) -> resolved: object <a name="addSchema"></a>

Add a schema to a current djv environment,

```javascript
env.addSchema('test', jsonSchema);
/* => {
  fn: function f0(data){...}
  name: 'test'
  schema: ...
} */
```

### validate(name: string, object: object) -> error: string <a name="validate"></a>

Check if object is valid against the schema

```javascript
env.validate('test#/common', { type: 'common' });
// => undefined

env.validate('test#/common', { type: 'custom' });
// => 'required: data'
```

where

* *name* - schema path in current environment
* *object* - object to validate
* *error* - undefined if it is valid

### removeSchema(name: string) <a name="removeSchema"></a>

Remove a schema or the whole structure from the djv environment

```javascript
env.removeSchema('test');
```

### resolve(name: string?) <a name="resolve"></a>

Resolve the name by existing environment

```javascript
env.resolve('test');
// => { name: 'test', schema: {} }, fn: ... }
```

### export(name: string?) -> state: object <a name="export"></a>

Export the whole structure object from environment or resolved by a given name

```javascript
env.export();
// => { test: { name: 'test', schema: {}, ... } }
```

where **state** is an internal structure or only resolved schema object

### import(config: object) <a name="import"></a>

Import all found structure objects to internal environment structure

```javascript
env.import(config);
```

### addFormat(name: string, formatter: string/function) <a name="addFormat"></a>

Add formatter to djv environment. When a string is passed it is interpreted as an expression which when returns `true` goes with an error, when returns `false` then a property is valid. When a function is passed it will be executed during schema compilation with a current schema and template helper arguments.

```javascript
env.addFormat('UpperCase', '%s !== %s.toUpperCase()');
// or
env.addFormat('isOk', function(schema, tpl){
  return `!${schema.isOk}`;
});
env.validate('ok', 'valid') // => undefined if schema contains isOk property
```

### setErrorHandler(errorHandler: function) <a name="errorHandler"></a>

Specify custom error handler which will be used in generated functions when problem found.
The function should return a string expression, which will be executed when generated validator function is executed. The simplist use case is the default one @see `template/defaultErrorHandler`
```javascript
 function defaultErrorHandler(errorType) {
   return `return "${errorType}: ${tpl.data}";`;
 }
```
It returns an expression 'return ...', so the output is an error string.

```javascript
djv({ errorHandler: () => 'return { error: true };' }) // => returns an object
djv({
  errorHandler(type) {
    return `errors.push({
      type: '${type}',
      schema: '${this.schema[this.schema.length - 1]}',
      data: '${this.data[this.data.length - 1]}'
    });`;
  }
});
```
When a custom error handler is used, the template body function adds a `error` variable inside a generated validator, which can be used to put error information. `errorType` is always passed to error handler function. Some validate utilities put extra argument, like f.e. currently processed property value. Inside the handler context is a templater instance, which contains `this.schema`, `this.data` paths arrays to identify validator position.
@see test/index/setErrorHandler for more examples

### useVersion(version: string, configure: function) <a name="useVersion"></a>

To customize environment provide a `configure` function which will update configuration for *djv* instance.

```javascript
env.useVersion('draft-04')
// or
env.useVersion('custom', configure)
```

`Configure` will get internal properties as an argument. Check the [@korzio/djv-draft-04 code](https://github.com/korzio/-djv-draft-04/blob/master/index.js).

```javascript
exposed = {
  properties,
  keywords,
  validators,
  formats,
  keys,
  transformation,
}
```

**!Important** Modifying them will affect all *djv* instances in an application.

## Tests <a name="tests"></a>

```
npm test
```

## References <a name="references"></a>

* [JSON-Schema Specification](http://json-schema.org/)
* [JSON-Schema Benchmark](https://github.com/ebdrup/json-schema-benchmark)


CHANGELOG.md
#### 2.1.4 (2020-12-27)

##### Chores

* **npm:**  regenerate package lock ([46bfdbe8](https://github.com/korzio/djv/commit/46bfdbe86fe821e9bb4ed9b39ea1830840204742))
* **code:**  Move clear decode function to outer scope ([5e65fda4](https://github.com/korzio/djv/commit/5e65fda4edbaa7a7495fd759fb5ab942fa06308f))

#### 2.1.2 (2019-03-06)

##### Other Changes

*  Update package version to 2.1.2 [#76](https://github.com/korzio/djv/pull/76) ([9b40c9d5](https://github.com/korzio/djv/commit/9b40c9d5ae50ce0d1324b5b25d3d9f8b8169882e))
*  Fix required validator throwing on null values [#75](https://github.com/korzio/djv/pull/76)

#### 2.1.1 (2018-09-04)

##### Other Changes

*  Update package version to 2.1.1 [#69](https://github.com/korzio/djv/pull/69) ([44190e9d](https://github.com/korzio/djv/commit/44190e9d9fab723942d93cc34a36c22e85c0b11d))

[Add `browser` field](https://docs.npmjs.com/files/package.json#browser)

#### 2.1.0 (2018-06-24)

##### New Features

* **error code** [BREAKING CHANGE] Change default error handlers to output error objects instead of strings ([#54](https://github.com/korzio/djv/commit/014a38285ede3e26455bb07c02d1613541892897))

##### Other Changes

* Update package version to 2.1.0 [#65](https://github.com/korzio/djv/pull/65) ([6fd3445f](https://github.com/korzio/djv/commit/6fd3445f3a7374d186a2519f068826a76e0887ea))

#### 2.0.0 (2017-10-10)

##### New Features

* **env** [BREAKING CHANGE] Make schema v6 default #48 ([bd1d4cae](https://github.com/korzio/djv/commit/bd1d4cae9d275c188536d37d721236745c4a1dc2))

##### Documentation Changes

* **api** Add auto-generated docs #48 ([440ff67d](https://github.com/korzio/djv/commit/440ff67d84aad429ab5a334aed384f59c5a96aa7))
* **site** Add index documentation page #48 ([cb4db59b](https://github.com/korzio/djv/commit/cb4db59b1db66f5107a834b236ad461e04e3e6cb))
* **env** Add environment section to describe options ([5e0e1a91](https://github.com/korzio/djv/commit/5e0e1a91e1eead2942032545e220e44f24c87dfd))

#### 1.2.0 (2017-09-25)

##### Documentation Changes

* **npm:** Remove docs folder from npm files #51 ([d5d2a8de](https://github.com/korzio/djv/commit/d5d2a8decd75db3351734ba25175b98533ca2d75))
* **readme:** Update v6 support #51 ([2ec9d7f0](https://github.com/korzio/djv/commit/2ec9d7f0abe53edb5b0e0c2d23b0747dd1cafbca))
* **const:** Add const documentation #51 ([b7f0355d](https://github.com/korzio/djv/commit/b7f0355d61e048d88ee485007d006f47831de0c9))
* **state:**
  * Fix state documentation #51 ([7a000072](https://github.com/korzio/djv/commit/7a000072b038bf6ffa3a4d0c98252c739fecd7c3))
  * Add resolution exampes #41 ([a89658c0](https://github.com/korzio/djv/commit/a89658c07dfb65848920425404e278b5c099eaa7))
  * Add state description #41 ([ff1b7a75](https://github.com/korzio/djv/commit/ff1b7a75392e807098c29d0a719b99cdb1f596d6))
* **todo:** Add todo for draft-06 #41 ([4fa64971](https://github.com/korzio/djv/commit/4fa64971c550fa558fda3031c93bd62ef061a7aa))
* **contribute:** Change tests debug description ([b0d03069](https://github.com/korzio/djv/commit/b0d03069c24c462cee82c8c6b5c78baeb4cc4f41))
* **contributing:**
  * Fix version ([351a8eb8](https://github.com/korzio/djv/commit/351a8eb88648b2939ecc7e6a87c9c6ce1f762f1d))
  * Fix typo ([66b62f3e](https://github.com/korzio/djv/commit/66b62f3eab75768403eae9682bb49ebf859f99eb))
  * Add release branch MR description ([faf53856](https://github.com/korzio/djv/commit/faf538564894a47d386e621b7de8a21602020052))
* **release:** Add table of contents, update release contributing #35 ([2f69a166](https://github.com/korzio/djv/commit/2f69a1660c3d882bc783e42dbd8ea45e7173d824))
* Update api docs, format utils ([11cae181](https://github.com/korzio/djv/commit/11cae1815de6f786abc24a71616eb50e18b2beaf))

##### New Features

* **id:** Add $id, examples keywords #44 ([1ef4745f](https://github.com/korzio/djv/commit/1ef4745f1658857c6c132264b8319fdaa73ccbcc))
* **format:** Add uri formats #44 ([79802269](https://github.com/korzio/djv/commit/79802269650e530228f2719497ddda165acfba81))
* **validator:**
  * Add json-pointer format #44 ([7a522f8a](https://github.com/korzio/djv/commit/7a522f8a9da29504190140329d12faf0aa29d4c5))
  * Add propertyNames validator #44 ([5ad9d343](https://github.com/korzio/djv/commit/5ad9d343575558be0e6a69bd2420b958e1d4dc00))
  * Add const validator #44 ([1d4fa106](https://github.com/korzio/djv/commit/1d4fa10656d4bb7caffd2fd9b598ecfccf1d71eb))
  * Add contains validator #44 ([8ecf27d8](https://github.com/korzio/djv/commit/8ecf27d81621b7638e6199c75c1ab623f47561ec))
* **validators:** Add useVersion method to change environment version; Add exlusive maximum/minimum properties #44 ([f57f9116](https://github.com/korzio/djv/commit/f57f9116204ad0f9f4126a81e6ead89bbce25f58))

##### Bug Fixes

* **test:**
  * Add name json mock data to test environment #41 ([c1d0e13b](https://github.com/korzio/djv/commit/c1d0e13bb50ccab189e88b7f6d6e2568c47a3770))
* **rename:** Rename template resolve to link as in state #41 ([34fc6e4d](https://github.com/korzio/djv/commit/34fc6e4d3f378a82378244a354817cd9ec4a6e3e))
* **format:** Add regex, fix uri formatters #44 ([757397d7](https://github.com/korzio/djv/commit/757397d7fb1a698a92953596804cf37d28b325d1))
* **draft-06:**
  * Move validators to patch environment #44 ([9f7412bc](https://github.com/korzio/djv/commit/9f7412bc3ed8148d119b729cf30b6f8d4c7fbb82))
  * Fix properties strict type check #44 ([f663f7ba](https://github.com/korzio/djv/commit/f663f7ba6285ed04575050ac9d27eb4509420c01))
* **boolean:** Fix transform default schema ([69488621](https://github.com/korzio/djv/commit/694886218a5f1330d5c49671d88932b645185b95))
* **tests:** Add isSchema, transformSchema utilities for boolean schema presentation #44 ([cb48a2a1](https://github.com/korzio/djv/commit/cb48a2a1da27a7a9c1c373a412b2c0624b8f8f7d))
* **validator:** Add not array required check #44 ([83b01279](https://github.com/korzio/djv/commit/83b01279824698d783aedc8c39b0b434c404f22b))

##### Other Changes


src/djv.ts
import { restore, expression } from './utils/template';
import formats from './utils/formats';
import { generate, State } from './utils/state';
import { add, use } from './utils/environment.js';

/**
 * Configuration for template
 * @typedef {object} DjvConfig
 * @property {string?} version - defines which version of json-schema draft to use,
 * draft-04 by default
 * @property {function?} versionConfigure - handler to apply for environment version
 * @property {boolean?} inner - a generating object should be considered as inner
 * Default value is false/undefined.
 * If true, then it avoid creating variables in a generated function body,
 * however without proper wrapper function approach will not work.
 * @see template/body, template/body
 * @property {object?} formats - an object containing list of formatters to add for environment
 * @property {function?} errorHandler - a handler to use for generating custom error messages
 */

/**
 * @name Environment
 * @description
 * Key constructor used for creating enivornment instance
 * @type {function} constructor
 * @param {DjvConfig} options passed to templater and utilities
 *
 * Usage
 *
 * ```javascript
 * const env = djv();
 * const env = new djv();
 * const env = new djv({ errorHandler: () => ';' });
 * ```
 */

export interface IOptions {
  version: string;
  versionConfigure?: Function;
  formats?: IFormats;
  inner: boolean;
  errorHandler?: Function;
}

interface IFormats {
  [key: string]: Function;
}

export interface ISchema {
  id: string;
  toString:(a: string) => string
}

interface IResolve {
  name: string;
  schema: any;
  fn: any;
}

type INoNameResolve = Omit<IResolve, 'name'>

class Environment {
  static expression = expression

  state: State
  options: IOptions;
  private defaultOptions: IOptions = {
    version: 'draft-06',
    inner: false,
  }
  resolved: {
    [key: string]: IResolve
  };

  constructor(options: Partial<IOptions> = {}) {
    const enrichedOptions: IOptions = {
      ...this.defaultOptions,
      ...options
    }
    const {version, versionConfigure} = enrichedOptions;
    this.options = enrichedOptions;
    this.resolved = {};
    this.state = new State(null, this);

    this.useVersion(version, versionConfigure);
    formats.setFormatters(enrichedOptions.formats || {})

  }
  /**
   * check if object correspond to schema
   *
   * Usage
   *
   * ```javascript
   * env.validate('test#/common', { type: 'common' });
   * // => undefined
   *
   * env.validate('test#/common', { type: 'custom' });
   * // => 'required: data'
   *
   * @param {string} name
   * @param {object} object
   * @returns {string} error - undefined if it is valid
   */
  validate(name: string, object: object) {
    return this.resolve(name).fn(object);
  }
  /**
   * add schema to djv environment
   *
   * Usage
   *
   * ```javascript
   * env.addSchema('test', jsonSchema);
   * ```
   *
   * @param {string?} name
   * @param {object} schema
   * @returns {resolved}
   */
  addSchema(name: string, schema: ISchema): INoNameResolve {
    const realSchema = typeof name === 'object' ? name : schema;
    const resolved = {
      schema: realSchema,
      fn: generate(this, realSchema, undefined, this.options),
    };

    [name, schema.id]
        // .filter(id => typeof id === 'string')
        .forEach((id) => {
          this.resolved[id] = Object.assign({ name: id }, resolved);
        });

    return resolved;
  }
  /**
   * removes a schema or the whole structure from djv environment
   *
   * Usage
   *
   * ```javascript
   * env.removeSchema('test');
   * ```
   *
   * @param {string} name
   */
  removeSchema(name: string) {
    if (name) {
      delete this.resolved[name];
    } else {
      this.resolved = {};
    }
  }
  /**
   * resolves name by existing environment
   *
   * Usage
   *
   * ```javascript
   * env.resolve('test');
   * // => { name: 'test', schema: {} }, fn: ... }
   * ```
   *
   * @param {string} name
   * @returns {resolved}
   */
  resolve(name: string): INoNameResolve {
    if (typeof name === 'object' || !this.resolved[name]) {
      return this.addSchema(
          name,
          this.state.resolve(name)
      );
    }

    return this.resolved[name];
  }
  /**
   * exports the whole structure object from environment or by resolved name
   *
   * Usage
   *
   * ```javascript
   * env.export();
   * // => { test: { name: 'test', schema: {}, ... } }
   * ```
   *
   * @param {string} name
   * @returns {serializedInternalState}
   */
  export(name: string) {
    let resolved: any;
    if (name) {
      const resolvedNoName = this.resolve(name);
      resolved = {
        name,
        schema: resolvedNoName.schema,
        fn: resolvedNoName.fn.toString()
      };
    } else {
      resolved = {};
      Object.keys(this.resolved).forEach((key) => {
        resolved[key] = {
          name: key,
          schema: this.resolved[key].schema,
          fn: this.resolved[key].fn.toString()
        };
      });
    }

    return JSON.stringify(resolved);
  }
  /**
   * imports all found structure objects to internal environment structure
   * Usage
   *
   * ```javascript
   * env.import(config);
   * ```
   *
   * @param {object} config - internal structure or only resolved schema object
   */
  import(config: string) {
    const item = JSON.parse(config);
    let restoreData = item;
    if (item.name && item.fn && item.schema) {
      restoreData = { [item.name]: item };
    }

    Object.keys(restoreData).forEach((key) => {
      const { name, schema, fn: source } = restoreData[key];
      const fn = restore(source, schema, this.options);
      this.resolved[name] = { name, schema, fn };
    });
  }
  /**
   * @name addFormat
   * @type function
   * @description
   * Add formatter to djv environment.
   * When a string is passed it is interpreted as an expression which
   * when returns `true` goes with an error, when returns `false` then a property is valid.
   * When a function is passed it will be executed during schema compilation
   * with a current schema and template helper arguments.
   * @see utils/formats
   *
   * Usage
   *
   * ```javascript
   * env.addFormat('UpperCase', '%s !== %s.toUpperCase()');
   * // or
   * env.addFormat('isOk', function(schema, tpl){
   *   return `!${schema.isOk} || %s !== %s.toUpperCase()`;
   * });
   * ```
   *
   * @param {string/object?} name
   * @param {string/function} formatter
   */


  //
  // addFormat(name: string | IFormats, formatter?: Function) {
  //   if (typeof name === 'string' && formatter) {
  //     formats.setFormatter(name, formatter)
  //     return;
  //   }
  //
  //   if (typeof name === 'object') {
  //     formats.setFormatters()
  //     Object.assign<IFormats, IFormats>(formats, name);
  //   }
  // }



  /**
   * @name setErrorHandler
   * @type function
   * @description
   * Specify custom error handler which will be used in generated functions when problem found.
   * The function should return a string expression, which will be executed when generated
   * validator function is executed. The simpliest use case is the default one
   * @see template/defaultErrorHandler
   * ```javascript
   *  function defaultErrorHandler(errorType) {
   *    return `return "${errorType}: ${tpl.data}";`;
   *  }
   * ```
   * It returns an expression 'return ...', so the output is an error string.
   * Usage
   * ```javascript
   * djv({ errorHandler: () => 'return { error: true };' }) // => returns an object
   * djv({
   *  errorHandler: function customErrorHandler(errorType, property) {
   *    return `errors.push({
   *      type: "${errorType}",
   *      schema: "${this.schema[this.schema.length - 1]}",
   *      data: "${this.data[this.data.length - 1]}"
   *    });`;
   *  }
   * });
   * ```
   * When a custom error handler is used, the template body function adds a `error` variable inside
   * a generated validator, which can be used to put error information. `errorType` is always
   * passed to error handler function. Some validate utilities put extra argument, like f.e.
   * currently processed property value. Inside the handler context is a templater instance,
   * which contains `this.schema`, `this.data` paths arrays to identify validator position.
   * @see test/index/setErrorHandler for more examples
   * @param {function} errorHandler - a function called each time compiler creates an error branch
   * @returns void
   */
  setErrorHandler(errorHandler: Function) {
    Object.assign(this.options, { errorHandler });
  }
  /**
   * @name useVersion
   * @type {function}
   * @description
   * Add a specification version for environment
   * A configure function is called with exposed environments, like keys, formats, etc.
   * Updates internals utilities and configurations to fix versions implementation conflicts
   * @param {string} version of json-schema specification to use
   * @param {function} configure
   * @returns void
   */
  useVersion(version: string, configure?: Function) {
    if (typeof configure !== 'function' && version === 'draft-04') {
      /* eslint-disable no-param-reassign, global-require, import/no-extraneous-dependencies */
      configure = require('@korzio/djv-draft-04');
      /* eslint-enable no-param-reassign, global-require, import/no-extraneous-dependencies */
    }
    if (typeof configure === 'function') {
      add(version, configure);
    }
    use(version);
  }
}

export default Environment;
